src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs

```diff
@@ -129,6 +141,42 @@ public ActionListExecutor(ILogger<CollectionRuleService> logger, ICollectionRule
             }
         }
 
+        private async Task WaitForCompletion(CollectionRuleContext context,
+            Action startCallback,
+            ICollectionRuleAction action,
+            CollectionRuleActionOptions actionOption,
+            CancellationToken cancellationToken)
+        {
+            //Before we wait for any other action to complete, we signal that execution has started.
+            //This allows process resume to occur.
+            startCallback?.Invoke();
+
+            CollectionRuleActionResult results = await action.WaitForCompletionAsync(cancellationToken);
+            if (!string.IsNullOrEmpty(actionOption.Name))
+            {
+                context.ActionResults.Add(actionOption.Name, results);
+            }
+
+            _logger.CollectionRuleActionCompleted(context.Name, actionOption.Type);
+        }
+
+        private async Task WaitForCompletion(CollectionRuleContext context,
+            Action startCallback, ActionCompletionEntry entry, CancellationToken cancellationToken)
+        {
+            try
```

Comment:
The try...catch...finally should be put into the other WaitForCompletion method so that end of the ExecuteActions method that waits for all of the remaining completes can handle exceptions, log them, and wrap them as necessary.